### PR TITLE
[Spec] Enhancing Java declarations with annotations

### PR DESCRIPTION
**Pull request by abreslav**

### CODE REVIEW

**Comment by abreslav** [comment]
@erokhins we'll try discuss specs as PRs

File: spec-docs/flexible-java-types.md [hunk 1]
```diff
@@ -139,4 +139,214 @@ The compiler issues warnings specific to `@Nullable`/`@NotNull` in the following
...
+Goals:
+ - Catch more errors related to nullability in case of Java interop
+ - Keep all class hierarchies consistent at all times (no hierarchy errors related to incorrect annotations)
+ - (!) If the code compiled with annotations, it should always compile without annotations (because external annotations may disappear)
+
+ This process never results in errors. On any mismatch, a bare platform signature is used (and a warning issued).
+
+### Annotations recognized by the compiler
+
+- `org.jetbrains.annotations.Nullable` - value may be null/accepts nulls
```
Comment (mikehearn) [inline spec-docs/flexible-java-types.md]: Why only the JetBrains versions and not also javax.annotations versions? My own code won't benefit if this is the complete list.

File: spec-docs/flexible-java-types.md [hunk 2]
```diff
@@ -139,4 +139,214 @@ The compiler issues warnings specific to `@Nullable`/`@NotNull` in the following
...
+- `org.jetbrains.annotations.NotNull` - value can not be null/passing null leads to an exception
+- `org.jetbrains.annotations.ReadOnly` - only non-mutating methods can be used on this collection/iterable/iterator
+- `org.jetbrains.annotations.Mutable` - only non-mutating methods can be used on this collection/iterable/iterator
```
Comment (mikehearn) [inline spec-docs/flexible-java-types.md]: This description seems duplicated?

Comment (VladimirReshetnikov) [inline spec-docs/flexible-java-types.md]: Yes, looks like a copy-paste mistake.

File: spec-docs/flexible-java-types.md (hunk 1 above)
Comment (abreslav) [inline spec-docs/flexible-java-types.md]: This may be extensible, but there are semantic differences which we need to look into

Comment (abreslav) [inline spec-docs/flexible-java-types.md]: We could also support the ones from [Android](https://developer.android.com/reference/android/support/annotation/NonNull.html) and [FindBugs](http://findbugs.sourceforge.net/manual/annotations.html) by default

Comment (mikehearn) [inline spec-docs/flexible-java-types.md]: Is there some subtle difference in the meaning of Nullable in the javax.annotations package that I'm not aware of? How many definitions can there be?

Comment (abreslav) [inline spec-docs/flexible-java-types.md]: I re-checked, and those differences (mostly in what NotNull means) seem to be irrelevant, so I added a list of other annotation packages below

Comment (abreslav) [inline spec-docs/flexible-java-types.md]: Thanks for your feedback!
